### PR TITLE
Creation of initial Specification and Conditions prototype

### PR DESCRIPTION
**Pull request by alelom**
### NOTE: Depends on
https://github.com/BHoM/BHoM/pull/1165

### Description
Initial implementation of the Specification and Conditions.

### Test file
https://mottmac.sharepoint.com/:f:/r/sites/PlatformDesignProgramme/Shared%20Documents/0.10%20Digital%20Configurator/WP5%20Reference%20Implementation/Scripts/%231-InitialImplementation?csf=1&web=1&e=hgCMsd

image

image

### CODE REVIEW

**Comment by al-fisher** [comment]
@alelom This is looking good - will have a bit more of a play with it tomorrow. 😸 

Initial comment around the handling of the Boolean operators -

Structured so as to accept a `List<IFilter>` and an `enum` in this way does leave the door open to some ambiguity for some of the operators.
For instance an `IMPLIES` must only have two arguments and is not a commutative operator, and so both the number and order of the inputs matters. You can create the below which is not clear

image



might therefore be easier to design as discrete objects and properties? What do you think?

**Comment by al-fisher** [comment]
Thinking further on the formatting of above _could_ be good to also think about structure of concepts for reflection into Excel_Toolkit! 

Building specs in tabular format through Excel templates will be very powerful.
Noting here - but perhaps one for later iterations

**Comment by alelom** [comment]
> For instance an IMPLIES must only have two arguments and is not a commutative operator, and so both the number and order of the inputs matters. You can create the below which is not clear

I see what you mean, but I think it is not different from any other operator, as they are all defined for 2 inputs only. The IMPLIES may simply work so if more than 2 inputs are defined e.g. A, B, C, D they resolve as [(A => B) => C] => D. Only difference as you say is that they are commutative, and in this the list order is important – however the user has still full control. Plus, it's less common as an operator, and I included it only for completeness. Happy to chat.

**Comment by alelom** [comment]
While implementing checks: the concept of as in our [Domain class](https://github.com/BHoM/BHoM/blob/ba4e0bddcf8bb86497b2e10f1663ab2f31319217/Data_oM/Collections/Domain.cs) may be too restrictive. Common scenario can be a Domain check for e.g. DateTime. I'm implementing it with System.Object. Only requirement for the check to work is that the type implements the `>`, `<`, `==` operators (as DateTime does). This way it is extendable to any measurable quantity. Happy to discuss.

File: CIH_Engine/CIH_Engine.csproj [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="BHoM">
+      <HintPath>..\..\..\..\..\ProgramData\BHoM\Assemblies\BHoM.dll</HintPath>
```
Comment (al-fisher) [inline CIH_Engine/CIH_Engine.csproj:35]:
```suggestion
      <HintPath>C:\ProgramData\BHoM\Assemblies\BHoM.dll</HintPath>
```

File: CIH_Engine/CIH_Engine.csproj [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    </Reference>
+    <Reference Include="Data_oM, Version=4.0.0.0, Culture=neutral, PublicKeyToken=null" />
+    <Reference Include="Reflection_Engine">
+      <HintPath>..\..\BHoM_Engine\Build\Reflection_Engine.dll</HintPath>
```
Comment (al-fisher) [inline CIH_Engine/CIH_Engine.csproj:39]:
```suggestion
      <HintPath>C:\ProgramData\BHoM\Assemblies\Reflection_Engine.dll</HintPath>
```

**Comment by al-fisher** [comment]
@alelom Added a couple of comments 👆 👆  in the absence of BHoMBot 🤖 
😄

File: CIH_oM/Conditions/ValueInSet.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+using BH.oM.Data.Library;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace BH.oM.Data.Conditions
+{
+    public class ValueInSet : BaseCondition 
+    {
+        List<object> Set { get; set; }
```
Comment (al-fisher) [inline CIH_oM/Conditions/ValueInSet.cs:35]:
```suggestion
        public virtual List<object> Set { get; set; }
```

@alelom missing `public virtual` here

**Review by al-fisher — CHANGES_REQUESTED** [review]
Hey @alelom - this is looking really good! Great stuff

I have a few comments will capture as I think of them ... I am continuing to experiment with and test

First comment is around the structuring of the nested `Conditions` (`Value`/`ValueInDomain`/_etc.)  being owned as a discrete property of the `PropertyCondition`

image

This leads to the above - which I think as a couple of issues for the user. 

- The `PropertyName` the Condition applies to should I think be defined "Flat" in parallel with the actual `Value`, `Tolerance`, `Domain` etc. Think this is a more logical structure for the user - see original prototype in issue here: https://github.com/CIHDigitalConfigurator/CIH_Toolkit/issues/1
We don't want the two separating. 
- Additionally having it structured as is currently you end up with the Parent `Source` and `Comment` being duplicated at both levels as your above example.

This structure comes from, I presume, wanting to reuse the Conditions across the `PropertyCondition`, `CustomDataCondition`, and `FragmentCondition` right?

I might simplify this by putting the `PropertyName` on the Value/Domain etc. Conditions themselves as a simple `string` property to start with. This help link more directly the Reference Value to the PropertyName etc. 
This simple approach will work for both Properties and Custom Data anyway. Is just Fragments that are different. Which is worth perhaps a discussion if they need more thought

Happy to chat on the above - will continue testing

**Comment by alelom** [comment]
> The PropertyName the Condition applies to should I think be defined "Flat" in parallel with the actual Value, Tolerance, Domain etc. Think this is a more logical structure for the user - see original prototype in issue here: #1
We don't want the two separating.

That was actually my first prototype. I then moved away from that because, in order to cover all _"comparisons"_ (`Value`, `Domain` and `Set`) you end up having:
- 3"CustomData" conditions objects: `CustomDataValueCondition`, `CustomDataDomainCondition`, `CustomDataSetCondition`
- 3"Fragments" conditions objects, like the above,
- 3"Property" conditions objects, like the above.

Therefore, I refactored, thinking that this approach:
- Makes the life actually easier for the user (less components to search; easier to switch)
- Makes things more scalable, as I thought that more conditions where those "comparisons" could be attached to may appear in the future. 

However, I did realise that, as you say, there is repetition of Source and Comment at the two levels. 
To solve that, I had in mind to perhaps qualify `ValueCondition`, `DomainCondition` and `SetCondition` as "Comparisons" instead (`ValueComparison`, `DomainComparison`, `SetComparison`) – objects that acquire value only when attached to a parent condition.

**Comment by al-fisher** [comment]
Hey @alelom - just capturing our conversation earlier.

Suggested moving back to list of simple Conditions
`ValueCondition`
`DomainCondition`
`SetCondition`
etc.

All the above then have as a defining Property: `string PropertyName` that the Condition applies to.

This PropertyName can refer to an object's _Defining Property_, _Derived Property_, _Custom Data_, or _Property of a Fragment_ it does not matter. This is all handled in your GetValue method.

This negates the need for `PropertyCondition`, `CustomDataCondition`, `FragmentCondition` layer


Also means we can neatly implement exposure of Reflection of Engine methods for applying Conditions to Properties like `Length()` etc. very intuitively for users.

**Comment by al-fisher** [comment]
Think above captures it? 

Can also catch up and finish our chat on Specifications owning Specifications as needed ... 😄

File: CIH_Engine/Compute/ApplyCondition/ApplyDomainCondition.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                result.Pattern.Add(passed);
+            }
+
+            result.FailInfo = failInfo;
+            return result;
+        }
+
+        public static bool NumberInDomain(double number, BH.oM.Data.Collections.Domain domain, double tolerance)
+        {
+            if (true) // TODO: implement extremes.
```
Comment (alelom) [inline CIH_Engine/Compute/ApplyCondition/ApplyDomainCondition.cs:55]: To be done. Quick feature for another PR. We need to merge https://github.com/BHoM/BHoM/pull/1165 for this.

File: CIH_Engine/Compute/ApplyCondition/ApplyValueCondition.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                                passed = numericalValue < referenceNumValue + numTolerance;
+                                break;
+                            case (ValueComparisons.SmallerThanOrEqualTo):
+                                passed = numericalValue <= referenceNumValue + numTolerance;
+                                break;
+                            case (ValueComparisons.LargerThanOrEqualTo):
+                                passed = numericalValue >= referenceNumValue + numTolerance;
+                                break;
+                            case (ValueComparisons.LargerThan):
+                                passed = numericalValue > referenceNumValue + numTolerance;
```
Comment (al-fisher) [inline CIH_Engine/Compute/ApplyCondition/ApplyValueCondition.cs:63]:
```suggestion
                                passed = numericalValue >= referenceNumValue - numTolerance;
                                break;
                            case (ValueComparisons.LargerThan):
                                passed = numericalValue > referenceNumValue - numTolerance;
```

**Review by al-fisher — CHANGES_REQUESTED** [review]
Hi @alelom 
Have tested and workflow is I think now looking good.

A couple of issues - some I think we need to look at here on this PR together before merge. Others are just questions/suggestions that can be picked up later.

- Issue with tolerance being applied correctly for `GreaterThan` cases - see suggest commit for application if you agree

- Value Check IsEqual not working as expected for Strings
image

- Properties in CustomData are not being compared properly it seems
image

Examples above saved in folder along with your example

File: CIH_Engine/Compute/ApplyCondition/ApplyValueCondition.cs (hunk 5 above)
Comment (al-fisher) [reply CIH_Engine/Compute/ApplyCondition/ApplyValueCondition.cs:63]: Suggested commit above

**Comment by alelom** [comment]
@al-fisher thanks for spotting those, all fixed now.

**Comment by al-fisher** [comment]
Brilliant - thanks

**Review by al-fisher — APPROVED** [review]
Thanks @alelom 

This is looking great as first prototype now. Will approve and merge to continue to dev as new issues 

Cheers

**Pull request merged**